test/shared/src/main/scala-2.12/zio/test/AssertionVariants.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2019-2024 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+import zio.stacktracer.TracingImplicits.disableAutoTrace
+import zio.test.Assertion.Arguments.valueArgument
+import zio.test.{ErrorMessage => M}
+
+trait AssertionVariants {
+  private def diffProduct[T](
+    obj1: T,
+    obj2: T,
+    paramNames: List[String] = Nil,
+    rootClassName: Option[String] = None
+  ): String = {
+    val currClassName = rootClassName.getOrElse(obj1.getClass.getSimpleName)
+
+    (obj1, obj2) match {
+      case (seq1: Iterable[Any], seq2: Iterable[Any]) => {
+        val maxSize    = math.max(seq1.size, seq2.size)
+        val paddedSeq1 = seq1.toVector.padTo(maxSize, null)
+        val paddedSeq2 = seq2.toVector.padTo(maxSize, null)
+
+        paddedSeq1
+          .zip(paddedSeq2)
+          .zipWithIndex
+          .flatMap { case ((subObj1, subObj2), index) =>
+            val newParamName = s"[$index]"
+
+            if (subObj1 != subObj2 && !subObj1.isInstanceOf[Product]) {
+              val paramName = s"${paramNames.reverse.mkString("")}[$index]"
+              Some(s"$currClassName$paramName : expected '$subObj2' got '$subObj1'\n")
+            } else {
+              diffProduct(subObj1, subObj2, newParamName :: paramNames, Some(currClassName))
+            }
+          }
+          .mkString
+      }
+      case (obj1: Product, obj2: Product) if obj1.productArity == obj2.productArity =>
+        obj1.productIterator
+          .zip(obj2.productIterator)
+          .zip(obj1.getClass.getDeclaredFields.iterator.map(_.getName))
```

Comment:
@urbit-pilled it looks like `getDeclaredFields` is not working with ScalaJS. Unfortunately the ZIO CI only runs ScalaJS tests with 2.13 so this was not caught before the release but this is showing up in other repos like https://github.com/zio/zio-query/actions/runs/9015988096/job/24771683500?pr=484

Any idea for a workaround?

Reply:
Sorry for the late response, I've been offline this week. Looks like #8841 fixes the ScalaJS and Scala Native issue?